presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1115,14 +1117,14 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
                 ImmutableList.of());
     }
 
-    private static TupleDomain<ColumnHandle> createPredicate(List<ColumnHandle> partitionColumns, List<HivePartition> partitions)
+    @VisibleForTesting
+    static TupleDomain<ColumnHandle> createPredicate(List<ColumnHandle> partitionColumns, List<HivePartition> partitions)
     {
         if (partitions.isEmpty()) {
             return TupleDomain.none();
         }
 
-        Map<ColumnHandle, Domain> domains = new HashMap<>(partitionColumns.size());
-
+        Multimap<ColumnHandle, NullableValue> partitionToValues = HashMultimap.create();
```

Comment:
You can do this whole thing in one shot, without having to pivot the column values into that intermediate multimap:

```java
static TupleDomain<ColumnHandle> createPredicate(List<ColumnHandle> partitionColumns, List<HivePartition> partitions)
{
    if (partitions.isEmpty()) {
        return TupleDomain.none();
    }

    return withColumnDomains(
            partitionColumns.stream()
                    .collect(Collectors.toMap(
                            Function.identity(),
                            column -> buildColumnDomain(partitions, column))));
}

private static Domain buildColumnDomain(List<HivePartition> partitions, ColumnHandle column)
{
    boolean hasNull = false;
    List<Object> nonNullValues = new ArrayList<>();
    Type type = null;

    for (HivePartition partition : partitions) {
        NullableValue value = partition.getKeys().get(column);
        if (value == null) {
            throw new PrestoException(HIVE_UNKNOWN_ERROR, format("Partition %s does not have a value for partition column %s", partition, column));
        }

        if (value.isNull()) {
            hasNull = true;
        }
        else {
            nonNullValues.add(value.getValue());
        }

        if (type == null) {
            type = value.getType();
        }
    }

    Domain domain = Domain.multipleValues(type, nonNullValues.toArray());

    if (hasNull) {
        domain = domain.union(Domain.onlyNull(type));
    }

    return domain;
}
```